<a href="https://colab.research.google.com/github/Abdullah-NI/NoteBooks/blob/main/30_GridsearchCV_and_RandomizedSearchCV.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Hyperparameter Tuning:**

1. GridSearchCV

2. RandomizedSearchCV

In [1]:
import pandas as pd
import numpy as np
import sklearn.datasets
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV

We will be working on the breast cancer dataset

In [2]:
breast_cancer_dataset=sklearn.datasets.load_breast_cancer()

In [3]:
print(breast_cancer_dataset)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

In [4]:
data_frame=pd.DataFrame(breast_cancer_dataset.data,columns=breast_cancer_dataset.feature_names)

In [5]:
data_frame.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [6]:
data_frame['label']=breast_cancer_dataset.target

In [7]:
data_frame.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  label  
0          0.4601                  0.11890      0  
1          0.2750                  0.08902      0  
2          0.3613                  0.08758      0  
3          0.6638                  0.17300      0  
4          0.2364                  0.07678      0  

[5 rows x 31 columns]

In [8]:
data_frame.shape

(569, 31)

In [9]:
data_frame.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
label                      0
dtype: int64

In [10]:
data_frame['label'].value_counts()

label
1    357
0    212
Name: count, dtype: int64

1 --> Benign

0 --> Malignant

In [11]:
x=data_frame.drop(columns='label',axis=1)
y=data_frame['label']

In [12]:
print(x)

     mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mea

In [13]:
print(y)

0      0
1      0
2      0
3      0
4      0
      ..
564    0
565    0
566    0
567    0
568    1
Name: label, Length: 569, dtype: int64


In [14]:
x=np.asarray(x)
y=np.asarray(y)

In [15]:
print(x)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]


In [16]:
print(y)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 0 1 1 1 1 0 0 0 1 1
 1 1 0 1 0 1 0 1 1 1 0 1 1 1 1 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0
 0 1 0 0 1 1 1 1 1 0 1 1 1 1 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1
 1 0 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 1 1
 0 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1
 1 1 1 1 1 1 0 1 0 1 1 0 

#**<h1>Gridsearch cv**

GridSearchCV is used for determining the best parameters for our model

In [17]:
model=SVC()

In [18]:
#hyperparameter dictionary
parameters={
    'kernel':['linear','poly','rbf','sigmoid'],
    'C':[1,5,10,20]
}

In [19]:
#grid search
classifier=GridSearchCV(model,parameters,cv=5)  # 5 part banakr somthing like that


In [20]:
#fittting data to our model
classifier.fit(x,y)

GridSearchCV(cv=5, estimator=SVC(),
             param_grid={'C': [1, 5, 10, 20],
                         'kernel': ['linear', 'poly', 'rbf', 'sigmoid']})

In [21]:
classifier.cv_results_

{'mean_fit_time': array([1.69696450e+00, 6.76569939e-03, 8.00156593e-03, 2.36350536e-02,
        3.22384677e+00, 4.16917801e-03, 5.98368645e-03, 1.37047768e-02,
        4.65048699e+00, 7.37514496e-03, 6.71663284e-03, 2.21539021e-02,
        7.54464507e+00, 4.48546410e-03, 4.19468880e-03, 1.35865211e-02]),
 'std_fit_time': array([6.16169522e-01, 7.44268687e-04, 5.28319648e-05, 4.92589023e-04,
        8.10918424e-01, 3.61791828e-04, 1.54585513e-03, 5.25953677e-04,
        1.00208923e+00, 9.53405993e-04, 5.07394567e-04, 9.31731564e-04,
        2.56885768e+00, 2.36779568e-04, 1.34129673e-04, 6.37103821e-04]),
 'mean_score_time': array([0.00146003, 0.00252142, 0.00393319, 0.0060813 , 0.00139542,
        0.00130472, 0.00251489, 0.00361214, 0.00143661, 0.00195556,
        0.00281034, 0.00577989, 0.00132394, 0.00122008, 0.0016089 ,
        0.00375519]),
 'std_score_time': array([1.66974484e-04, 1.31071129e-04, 7.49111934e-04, 2.10770784e-04,
        2.31676200e-04, 4.96809859e-05, 6.52524453e-

In [22]:
# best parameter
best_parameter=classifier.best_params_
print(best_parameter)

{'C': 10, 'kernel': 'linear'}


In [23]:
#highest accuracy
highest_accuracy=classifier.best_score_
print(highest_accuracy)

0.9525694767893185


In [24]:
#loading result into pandas datafeme
result=pd.DataFrame(classifier.cv_results_)

In [25]:
result.head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time  param_C  \
0       1.696965      0.616170         0.001460        0.000167        1   
1       0.006766      0.000744         0.002521        0.000131        1   
2       0.008002      0.000053         0.003933        0.000749        1   
3       0.023635      0.000493         0.006081        0.000211        1   
4       3.223847      0.810918         0.001395        0.000232        5   

  param_kernel                         params  split0_test_score  \
0       linear   {'C': 1, 'kernel': 'linear'}           0.947368   
1         poly     {'C': 1, 'kernel': 'poly'}           0.842105   
2          rbf      {'C': 1, 'kernel': 'rbf'}           0.850877   
3      sigmoid  {'C': 1, 'kernel': 'sigmoid'}           0.543860   
4       linear   {'C': 5, 'kernel': 'linear'}           0.947368   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.929825           0.973684           0.921053           0.955752   
1           0.885965           0.929825           0.947368           0.938053   
2           0.894737           0.929825           0.947368           0.938053   
3           0.456140           0.464912           0.385965           0.451327   
4           0.938596           0.973684           0.929825           0.964602   

   mean_test_score  std_test_score  rank_test_score  
0         0.945536        0.018689                4  
1         0.908663        0.039382               12  
2         0.912172        0.035444               11  
3         0.460441        0.050253               13  
4         0.950815        0.016216                2

In [26]:
grid_search_result=result[['param_C','param_kernel','mean_test_score']]

In [27]:
grid_search_result

param_C param_kernel  mean_test_score
0         1       linear         0.945536
1         1         poly         0.908663
2         1          rbf         0.912172
3         1      sigmoid         0.460441
4         5       linear         0.950815
5         5         poly         0.922729
6         5          rbf         0.931501
7         5      sigmoid         0.411178
8        10       linear         0.952569
9        10         poly         0.920975
10       10          rbf         0.922714
11       10      sigmoid         0.402391
12       20       linear         0.949061
13       20         poly         0.919221
14       20          rbf         0.920944
15       20      sigmoid         0.398867

Highest Accuracy = 95.2%

Best Parameters = {'C':10, 'kernel':'linear'}

#**RandomizedSearchCV**

In [28]:
model=SVC()

In [29]:
#hyperparameter dictionary
parameters={
    'kernel':['linear','poly','rbf','sigmoid'],
    'C':[1,5,10,20]
}

In [30]:
classifier=RandomizedSearchCV(model,parameters,cv=5)

In [31]:
#fittting data to our model
classifier.fit(x,y)

RandomizedSearchCV(cv=5, estimator=SVC(),
                   param_distributions={'C': [1, 5, 10, 20],
                                        'kernel': ['linear', 'poly', 'rbf',
                                                   'sigmoid']})

In [32]:
classifier.cv_results_

{'mean_fit_time': array([0.01895933, 0.01147728, 0.00806208, 3.26602559, 0.00450659,
        0.00451603, 1.41042819, 0.02111521, 0.02159181, 0.00726748]),
 'std_fit_time': array([8.91072805e-03, 3.25784855e-03, 1.57258649e-03, 9.32813809e-01,
        2.48598539e-04, 1.23513398e-04, 5.39465726e-01, 1.35398900e-03,
        1.24967549e-03, 9.05391043e-04]),
 'mean_score_time': array([0.00927129, 0.00379372, 0.00328045, 0.00137982, 0.00120897,
        0.0018589 , 0.0013721 , 0.00531025, 0.00530577, 0.00297585]),
 'std_score_time': array([6.86149221e-03, 2.89207716e-03, 5.18336761e-04, 2.02725040e-04,
        3.64904756e-05, 6.75475474e-05, 2.18392928e-04, 1.62877718e-04,
        8.53451932e-05, 1.74647531e-04]),
 'param_kernel': masked_array(data=['rbf', 'poly', 'rbf', 'linear', 'poly', 'rbf',
                    'linear', 'sigmoid', 'sigmoid', 'rbf'],
              mask=[False, False, False, False, False, False, False, False,
                    False, False],
        fill_value='?',
    

In [33]:
# best parameter
best_parameter=classifier.best_params_
print(best_parameter)

{'kernel': 'linear', 'C': 5}


In [34]:
#highest accuracy
highest_accuracy=classifier.best_score_
print(highest_accuracy)

0.9508150908244062


In [35]:
#loading result into pandas datafeme
result=pd.DataFrame(classifier.cv_results_)

In [36]:
result.head()

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_kernel  \
0       0.018959      0.008911         0.009271        0.006861          rbf   
1       0.011477      0.003258         0.003794        0.002892         poly   
2       0.008062      0.001573         0.003280        0.000518          rbf   
3       3.266026      0.932814         0.001380        0.000203       linear   
4       0.004507      0.000249         0.001209        0.000036         poly   

   param_C                        params  split0_test_score  \
0       20    {'kernel': 'rbf', 'C': 20}           0.877193   
1        5    {'kernel': 'poly', 'C': 5}           0.885965   
2       10    {'kernel': 'rbf', 'C': 10}           0.877193   
3        5  {'kernel': 'linear', 'C': 5}           0.947368   
4       20   {'kernel': 'poly', 'C': 20}           0.877193   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.921053           0.921053           0.947368           0.938053   
1           0.912281           0.921053           0.938596           0.955752   
2           0.921053           0.912281           0.956140           0.946903   
3           0.938596           0.973684           0.929825           0.964602   
4           0.921053           0.903509           0.938596           0.955752   

   mean_test_score  std_test_score  rank_test_score  
0         0.920944        0.024105                6  
1         0.922729        0.023689                4  
2         0.922714        0.027879                5  
3         0.950815        0.016216                1  
4         0.919221        0.027300                7

In [37]:
grid_search_result=result[['param_C','param_kernel','mean_test_score']]

In [38]:
grid_search_result

param_C param_kernel  mean_test_score
0       20          rbf         0.920944
1        5         poly         0.922729
2       10          rbf         0.922714
3        5       linear         0.950815
4       20         poly         0.919221
5        1          rbf         0.912172
6        1       linear         0.945536
7       20      sigmoid         0.398867
8       10      sigmoid         0.402391
9        5          rbf         0.931501

Highest Accuracy = 95.08%

Best Parameters = {'C':5, 'kernel':'linear'}

**<h1>NOTE<h1/>**  :: now we get our hyperparameter so train the model with these parameter